File: go/tests/memory_encoder_test.go
// memory_encoder_test.go
//
// Verifies CompanionMemoryEncoder end-to-end: a turn handed to the background
// encoder fills the knowledge graph so associative recall can later reach the
// episode; attributed beliefs are formed off the hot path (a third party's fact
// never becomes the user's); the queue drops rather than blocks when full;
// Close drains remaining work; and an extractor failure is captured, not fatal.
// Mirrors the TS pilot suite tests/memory_encoder.test.ts 1:1.

package circleai_test

import (
	"context"
	"errors"
	"strings"
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
)

// throwingExtractor always errors from ExtractFromTurn.
type throwingExtractor struct{}

func (throwingExtractor) ExtractFromTurn(context.Context, string, string, *string) ([]circleai.KnowledgeTriple, error) {
	return nil, errors.New("boom")
}

func newEncoder(t *testing.T, ex circleai.IKnowledgeGraphExtractor, g *circleai.KnowledgeGraph, bx circleai.IBeliefExtractor, beliefs *circleai.SelfBeliefStore, capacity int) *circleai.CompanionMemoryEncoder {
	t.Helper()
	enc, err := circleai.NewCompanionMemoryEncoder(ex, g, bx, beliefs, capacity)
	if err != nil {
		t.Fatalf("NewCompanionMemoryEncoder: %v", err)
	}
	return enc
}

func TestMemoryEncoder_EndToEnd(t *testing.T) {
	ctx := context.Background()

	t.Run("encodes a turn so associative recall can reach the episode by a content word", func(t *testing.T) {
		graph := circleai.NewKnowledgeGraph()
		enc := newEncoder(t, &circleai.HeuristicKnowledgeGraphExtractor{}, graph, nil, nil, 0)

		enc.Enqueue("I love hiking in Drakensberg", "Sounds wonderful", "ep-hike")
		if err := enc.Close(); err != nil {
			t.Fatalf("Close: %v", err)
		}

		if len(graph.AllTriples()) == 0 {
			t.Fatalf("graph should have filled from the turn")
		}

		hippo := mustHippo(t, graph)
		hits, err := hippo.MultiHopRecall(ctx, "drakensberg", 5)
		if err != nil {
			t.Fatalf("MultiHopRecall: %v", err)
		}
		episode := findHit(hits, "ep-hike")
		if episode == nil {
			t.Fatalf("recall should reach the episode via the extracted edges")
		}
		if episode.Item.Text != "I love hiking in Drakensberg" {
			t.Errorf("episode text: got %q", episode.Item.Text)
		}
	})

	t.Run("forms attributed beliefs off the hot path — the mother's fact never becomes the user's", func(t *testing.T) {
		graph := circleai.NewKnowledgeGraph()
		beliefs := circleai.NewSelfBeliefStore()
		enc := newEncoder(t, &circleai.HeuristicKnowledgeGraphExtractor{}, graph, &circleai.HeuristicBeliefExtractor{}, beliefs, 0)

		enc.Enqueue("my mother is diabetic", "Noted", "ep1")
		enc.Enqueue("i am vegetarian", "Got it", "ep2")
		if err := enc.Close(); err != nil {
			t.Fatalf("Close: %v", err)
		}

		facts := beliefs.SelfFacts()
		for _, f := range facts {
			if strings.Contains(f.Object, "diabetic") {
				t.Errorf("mother's condition must never be a user fact")
			}
		}
		hasVeg := false
		for _, f := range facts {
			if f.Object == "vegetarian" {
				hasVeg = true
			}
		}
		if !hasVeg {
			t.Errorf("vegetarian should be a user fact")
		}
		if !nonSelfHas(beliefs, "diabetic") {
			t.Errorf("diabetic should still be remembered as an audit fact")
		}
	})
}

func TestMemoryEncoder_QueueBehaviour(t *testing.T) {
	t.Run("drops writes beyond capacity rather than blocking", func(t *testing.T) {
		graph := circleai.NewKnowledgeGraph()
		enc := newEncoder(t, &circleai.HeuristicKnowledgeGraphExtractor{}, graph, nil, nil, 2)

		// Enqueued before the drain is released (it drains on Close): the 3rd
		// overflows a capacity-2 queue and is dropped.
		enc.Enqueue("alpha", "", "e1")
		enc.Enqueue("bravo", "", "e2")
		enc.Enqueue("charlie", "", "e3")
		if err := enc.Close(); err != nil {
			t.Fatalf("Close: %v", err)
		}

		if graph.GetNode("e1") == nil {
			t.Errorf("e1 should be present")
		}
		if graph.GetNode("e2") == nil {
			t.Errorf("e2 should be present")
		}
		if graph.GetNode("e3") != nil {
			t.Errorf("the overflow write should have been dropped")
		}
	})

	t.Run("ignores an enqueue with a blank episode id", func(t *testing.T) {
		graph := circleai.NewKnowledgeGraph()
		enc := newEncoder(t, &circleai.HeuristicKnowledgeGraphExtractor{}, graph, nil, nil, 0)
		enc.Enqueue("hello", "", "")
		enc.Enqueue("hello", "", "   ")
		if err := enc.Close(); err != nil {
			t.Fatalf("Close: %v", err)
		}
		if len(graph.AllTriples()) != 0 {
			t.Errorf("blank episode ids should be ignored")
		}
	})

	t.Run("captures an extractor failure without crashing the drain", func(t *testing.T) {
		graph := circleai.NewKnowledgeGraph()
		enc := newEncoder(t, throwingExtractor{}, graph, nil, nil, 0)
		enc.Enqueue("x", "", "e1")
		if err := enc.Close(); err != nil {
			t.Fatalf("Close: %v", err)
		}

		last := enc.LastError()
		if last == nil || last.Error() != "boom" {
			t.Errorf("lastError: got %v want boom", last)
		}
		// The node was upserted before the extractor ran, so it survives.
		if graph.GetNode("e1") == nil {
			t.Errorf("e1 node should survive (upserted before the extractor ran)")
		}
	})
}

File: typescript/tests/memory_encoder.test.ts
// memory_encoder.test.ts
//
// Verifies CompanionMemoryEncoder end-to-end: a turn handed to the background
// encoder fills the knowledge graph so associative recall can later reach the
// episode; attributed beliefs are formed off the hot path (a third party's fact
// never becomes the user's); the queue drops rather than blocks when full;
// closeAsync drains remaining work; and an extractor failure is captured, not
// fatal.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { CompanionMemoryEncoder } from '../src/companion/memory_encoder';
import { HeuristicKnowledgeGraphExtractor, type IKnowledgeGraphExtractor } from '../src/memory/extractor';
import { HeuristicBeliefExtractor, SelfBeliefStore } from '../src/companion/belief';
import { InMemoryKnowledgeGraph, InMemoryHippoRagStore } from '../src/memory/graph';

describe('CompanionMemoryEncoder — end-to-end', () => {
  it('encodes a turn so associative recall can reach the episode by a content word', async () => {
    const graph = new InMemoryKnowledgeGraph();
    const enc = new CompanionMemoryEncoder(new HeuristicKnowledgeGraphExtractor(), graph);

    enc.enqueue('I love hiking in Drakensberg', 'Sounds wonderful', 'ep-hike');
    await enc.closeAsync();

    assert.ok(graph.allTriples().length > 0, 'graph should have filled from the turn');

    const hippo = new InMemoryHippoRagStore(graph);
    const hits = await hippo.multiHopRecallAsync('drakensberg', 5);
    const episode = hits.find((h) => h.item.id === 'ep-hike');
    assert.ok(episode, 'recall should reach the episode via the extracted edges');
    assert.equal(episode!.item.text, 'I love hiking in Drakensberg');
  });

  it('forms attributed beliefs off the hot path — the mother\'s fact never becomes the user\'s', async () => {
    const graph = new InMemoryKnowledgeGraph();
    const beliefs = new SelfBeliefStore();
    const enc = new CompanionMemoryEncoder(
      new HeuristicKnowledgeGraphExtractor(),
      graph,
      new HeuristicBeliefExtractor(),
      beliefs,
    );

    enc.enqueue('my mother is diabetic', 'Noted', 'ep1');
    enc.enqueue('i am vegetarian', 'Got it', 'ep2');
    await enc.closeAsync();

    const facts = beliefs.selfFacts();
    assert.ok(!facts.some((f) => f.object.includes('diabetic')), "mother's condition must never be a user fact");
    assert.ok(facts.some((f) => f.object === 'vegetarian'));
    assert.ok(beliefs.nonSelf().some((b) => b.object === 'diabetic'), 'it is still remembered as an audit fact');
  });
});

describe('CompanionMemoryEncoder — queue behaviour', () => {
  it('drops writes beyond capacity rather than blocking', async () => {
    const graph = new InMemoryKnowledgeGraph();
    const enc = new CompanionMemoryEncoder(new HeuristicKnowledgeGraphExtractor(), graph, null, null, 2);

    // Enqueued synchronously before the drain resumes: the 3rd overflows a
    // capacity-2 queue and is dropped.
    enc.enqueue('alpha', '', 'e1');
    enc.enqueue('bravo', '', 'e2');
    enc.enqueue('charlie', '', 'e3');
    await enc.closeAsync();

    assert.notEqual(graph.getNode('e1'), null);
    assert.notEqual(graph.getNode('e2'), null);
    assert.equal(graph.getNode('e3'), null, 'the overflow write should have been dropped');
  });

  it('ignores an enqueue with a blank episode id', async () => {
    const graph = new InMemoryKnowledgeGraph();
    const enc = new CompanionMemoryEncoder(new HeuristicKnowledgeGraphExtractor(), graph);
    enc.enqueue('hello', '', '');
    enc.enqueue('hello', '', '   ');
    await enc.closeAsync();
    assert.equal(graph.allTriples().length, 0);
  });

  it('captures an extractor failure without crashing the drain', async () => {
    const graph = new InMemoryKnowledgeGraph();
    const throwing: IKnowledgeGraphExtractor = {
      async extractFromTurnAsync() {
        throw new Error('boom');
      },
    };
    const enc = new CompanionMemoryEncoder(throwing, graph);
    enc.enqueue('x', '', 'e1');
    await enc.closeAsync();

    assert.ok(enc.lastError instanceof Error);
    assert.equal((enc.lastError as Error).message, 'boom');
    // The node was upserted before the extractor ran, so it survives.
    assert.notEqual(graph.getNode('e1'), null);
  });
});
